File: p_블록이동하기.py
from collections import deque

def solution(board):
    q = deque([(0, 0, 0, 1, 1)]) # 좌x, 좌y, 우x, 우y
    moves = [(-1, 0, -1, 0), (1, 0, 1, 0), (0, 0, -1, -1), (0, 0, 1, -1), (-1, 1, 0, 0), (1, 1, 0, 0),(0, 1, 0, 1),(0, -1, 0, -1)] # 총 6가지 이동 방법
    visit = set([(0, 0, 0, 1)])
    n = len(board)
    while q:
      a, b, c, d, cnt = q.popleft()
      if (a == n - 1 and b == n - 1) or (c == n - 1 and d == n - 1):

        return cnt

      for k in range(8):
        x, y, z, l = moves[k]
        a1 = a + x
        a2 = b + y
        b1 = c + z
        b2 = d + l


        if not (0 <= a1 < n and 0 <= a2 < n and 0 <= b1 < n and 0 <= b2 < n): # 범위 체크
          continue

        if not (0 <= board[a1][a2] < n and 0 <= board[b1][b2] < n):
          continue

        if board[a1][a2] == 1 or board[b1][b2] == 1: # 벽인지 확인
          continue
        
        # 원래 좌표 기준으로 대각선이 벽인지 확인하기
        if k == 2 and board[c-1][d] == 1: # 좌를 기준으로 위로 90도 회전한 경우
          continue

        elif k == 3 and board[c+1][d] == 1:
          continue

        elif k == 4 and board[a-1][b] == 1:
          continue

        elif k == 5 and board[a+1][b] == 1:
          continue
        
        elif (a1, a2, b1, b2) in visit: #방문했는지 체크
          continue

        q.append((a1, a2, b1, b2, cnt + 1)) #이동할 수 있는 경우 추가
        visit.add((a1, a2, b1, b2))



a = [[0, 0, 0, 1, 1],[0, 0, 0, 1, 0],[0, 1, 0, 1, 1],[1, 1, 0, 0, 1],[0, 0, 0, 0, 0]]
print(solution(a))

File: p_블록이동하기.js
function make_moves(a1, b1, a2, b2, board) {
  let candidates = [];
  // 평행 이동
  const upDown = [-1, 1]; // 상하
  upDown.forEach((item) => {
    if (board[a1 + item][b1] === 0 && board[a2 + item][b2] === 0) {
      candidates.push([a1 + item, b1, a2 + item, b2]);
    }
  });

  const leftRight = [-1, 1]; // 좌우
  leftRight.forEach((item) => {
    if (board[a1][b1 + item] === 0 && board[a2][b2 + item] === 0) {
      candidates.push([a1, b1 + item, a2, b2 + item]);
    }
  });

  if (a1 === a2) {
    // 가로 방향
    const garo = [-1, 1];
    garo.forEach((item) => {
      if (board[a1 + item][b1] === 0 && board[a2 + item][b2] === 0) {
        candidates.push([a1, b1, a1 + item, b1]);
        candidates.push([a2 + item, b2, a2, b2]);
      }
    });
  } else {
    const sero = [-1, 1];
    sero.forEach((item) => {
      if (board[a1][b1 + item] === 0 && board[a2][b2 + item] === 0) {
        candidates.push([a1, b1, a1, b1 + item]);
        candidates.push([a2, b2 + item, a2, b2]);
      }
    });
  }
  return candidates;
}

function solution(board) {
  const n = board.length;
  const new_board = new Array(n + 2)
    .fill()
    .map((_) => new Array(n + 2).fill(1));
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      new_board[i + 1][j + 1] = board[i][j];
    }
  }
  let queue = [[1, 1, 1, 2, 0]];
  let visit = new Set(["1112"]); // 값을 넣을 때는 배열 형태 아니고 문자열 형태로 할 때 시간초과 안났음
  while (queue.length) {
    const [x1, y1, x2, y2, cnt] = queue.shift();
    if ((x1 === n && y1 === n) || (x2 === n && y2 === n)) {
      return cnt;
    }
    make_moves(x1, y1, x2, y2, new_board).forEach((next) => {
      if (!visit.has(next.join(""))) {
        queue.push([...next, cnt + 1]);
        visit.add(next.join(""));
      }
    });
  }
}

let board = [
  [0, 0, 0, 1, 1],
  [0, 0, 0, 1, 0],
  [0, 1, 0, 1, 1],
  [1, 1, 0, 0, 1],
  [0, 0, 0, 0, 0],
];
console.log(solution(board));
